discussions/prog_repo.md

```diff
@@ -0,0 +1,192 @@
+can be downloaded to run on a variety of kernel versions, we propose that the repository follows
+best practices for compatibility, such as using eBPF CO-RE for Linux. Similar best practices
+can be followed on non-Linux platforms as they mature and become available.
+The user space components of eBPF programs pose a separate, complicated portability obstacle,
+which is compounded by the desire of the L3AF project to support user space programs in multiple
+languages. For the initial version of the repository, it should be sufficient to build
+(of a compiled language) and unit test for the x86_64 platform. The user space component
+should document any installation dependencies it has (e.g., MySQL, Grafana, Python libraries, etc.).
+Contributors to the repository would be responsible to provide the necessary build scripts
+and configuration.
+
+### Multiple Versions
+
+Regardless of whether the package is built from source or not, eBPF programs should use semantic
+versioning. The repository should then host previous versions (in addition to the current version)
+in some reasonable manner.
+
+### Automated Reviews
+
+There can be a mechanism to review source code, using tools to detect like code formatting errors,
+complexity, code issues, and code duplication. Similarly, vulnerability detection
+scanners can be used for artifacts before uploading.
+
+## Alternative to Hosting and Building Source Code
+
+An alternative to the L3AF Project hosting and building contributed source code would be for
+contributors to submit or self-host a signed package containing their eBPF programs and any
+documentation. The L3AF daemon, by default, would only download and run eBPF programs signed
+by trusted parties. This provides a nice segue for us to go into the signing section.
+
+## Signing and Verification
+
+There are three different layers of signing that should be discussed, the signing of the
+kernel eBPF program (i.e., byte code), the signing of the eBPF programs package (i.e., an archive
+containing the built eBPF program and any associated user space programs) and
+signing for Hypervisors (i.e., Virtualization-based Security).
+
+### Signing and verifying eBPF programs
+
+There are efforts underway in the eBPF community for how to sign and verify
+Linux eBPF programs, this enables JIT’ed native code:
+
+https://lore.kernel.org/bpf/20211203191844.69709-1-mcroce@linux.microsoft.com/
+https://linuxplumbersconf.org/event/11/contributions/947/
+
+Similarly, signing mechanism for other platforms are also being explored.
+
+L3AF can use the best practices that are established for signing and verifying eBPF programs.
+
+### Signing and verifying the eBPF Program package
+
+At a higher level, we also plan to sign and verify eBPF Package.
+The package can be signed by its creator using eBPF Package Repository generated/provided
+trusted keys. L3AF can, by default, verify the package comes from a trusted source before executing
+any programs in the package
```

Comment:
I think we would want to only allow l3af project reviewed and accepted programs to get signed with the eBPF repository trusted key. I would suggest l3af by default does not trust any 3rd party / trusted party signed packages and it would be required to configure l3afd to trust packages from a certain party.  For example, if Microsoft has signed eBPF packages, you should have to configure l3afd to explicitly allow signed packages from Microsoft.

Reply:
Agree, and to allow packages from the L3AF Public Repository (or whatever it's called) you should have to configure l3afd to explicitly allow signed packages from that repository.